### ISSUE TITLE
Printing from `tpplc`

### ISSUE THREAD

**Issue by vsenderov** [treeppl/treeppl#12]
Using `dprint` is not nice.

 ```mcore
  | PrintStmtTppl x ->
    lam cont.
      foldr isemi_ cont
        [ dprint_ (compileExprTppl x.printable)
        -- , print_ (str_ "\n") in
        , flushStdout_ unit_
        ]
```

---

Test program:

 ```tppl models/lang/hello.tppl
model function blub() {
  print("Hello world!");
}
```

```bash
build/tpplc models/lang/hello.tppl models/data/empty.mc out.mc
```

---

Output:

```
$ ./out 1 0
{ tag: 1, size: 1
, { tag: 0, size: 1
  , { tag: 0, size: 1
    , { tag: 0, size: 12 (as a string: Hello world!)
      , 72
      , 101
      , 108
      , 108
      , 111
      , 32
      , 119
      , 111
      , 114
      , 108
      , 100
      , 33
      }
    }
  }
}
```
--

## Objectives

1. What to substitute `drpint` with? We need to print _strings_, _reals_, _ints_, and _bools_, as well as other types that are supported.
2. Why does `out.mc 1 0` produce output?

## Analysis

It is possible to include Miking functions in TreePPL from the _data input file_.  If the function is an _intrinsic_ it will need to be wrapped in a let expression.

## Solution ideas:

- [x] the current TreePPL `print` is renamed to `debug`
- [x] we implement a `print` function in TreePPL specifically for strings
- [x] we need a TreePPL function that converts _Real_ to _String_ and (_Bool_, _Int_) (the beginnings of a standard library)

**Comment by vsenderov** [comment]
Side effects are always produced, even if you run 0 sweeps:

```
 build/tpplc models/lang/hello.tppl models/data/empty.mc out.mc && mi compile out.mc
viktor@yggy:~/Dropbox/Workspaces/Miking/treeppl_refactoring-models$ ./out 1 1
Hello, world!
Hello, world!
0.
() 0.
viktor@yggy:~/Dropbox/Workspaces/Miking/treeppl_refactoring-models$ ./out 1 1
Hello, world!
Hello, world!
0.
() 0.
viktor@yggy:~/Dropbox/Workspaces/Miking/treeppl_refactoring-models$ ./out 1 0
Hello, world!
viktor@yggy:~/Dropbox/Workspaces/Miking/treeppl_refactoring-models$ ./out 2 0
Hello, world!
```

**Comment by vsenderov** [comment]
Should we consider this a bug?

**Comment by vsenderov** [comment]
I am reopening #11 instead.

**Comment by vsenderov** [comment]
It seems that I have now a first draft ready in [[Refactoring models]]. 

- `print` function prints a string without a newline
- `printLn` prints a string with a newline
- `concat` concatenates strings together
- `int2string` converts an integer to string
- `real2string` converts a real number to string
- [ ] `bool2string` I have to implement

However, I wasn't able to implement a polymorphic function `tostr`. Lars gave me an idea:

```
type Num
con NumInt : Int -> Num
con NumFloat : Float -> Num
...
let x = NumInt 3
...
match x with NumInt y then
  ... (int)
else match x with NumFloat y then
  ... (float)
else
  ...
```

But if we do `let x = 3` then its type is not `Num`. I will make a new issue for that.

**Comment by vsenderov** [comment]
This issue ought to be closed if the PR #10 gets accepted and if the discussion does not prove otherwise.

**Comment by vsenderov** [comment]
merged
